File: packages/python/elizaos/features/advanced_capabilities/actions/follow_room.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class FollowRoomAction:
    name: str = "FOLLOW_ROOM"
    similes: list[str] = field(
        default_factory=lambda: [
            "JOIN_ROOM",
            "SUBSCRIBE_ROOM",
            "WATCH_ROOM",
            "ENTER_ROOM",
        ]
    )
    description: str = (
        "Follow a room to receive updates and monitor messages. "
        "Use this when you want to actively engage with a room's content."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None:
            return False

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                followed_rooms = world.metadata.get("followedRooms", [])
                if str(room_id) in followed_rooms:
                    return False

        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to follow",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "FOLLOW_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "FOLLOW_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                followed_rooms = list(world.metadata.get("followedRooms", []))
                room_id_str = str(room_id)

                if room_id_str not in followed_rooms:
                    followed_rooms.append(room_id_str)
                    world.metadata["followedRooms"] = followed_rooms
                    await runtime.update_world(world)

        await runtime.create_memory(
            content=Content(
                text=f"Now following room: {room_name}",
                actions=["FOLLOW_ROOM"],
            ),
            room_id=room_id,
            entity_id=runtime.agent_id,
            memory_type="action",
            metadata={"type": "FOLLOW_ROOM", "roomName": room_name},
        )

        response_content = Content(
            text=f"I am now following {room_name} and will monitor its messages.",
            actions=["FOLLOW_ROOM"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Now following room: {room_name}",
            values={
                "success": True,
                "following": True,
                "roomId": str(room_id),
                "roomName": room_name,
            },
            data={
                "actionName": "FOLLOW_ROOM",
                "roomId": str(room_id),
                "roomName": room_name,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Can you keep an eye on this channel?"),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll follow this room and monitor its activity.",
                        actions=["FOLLOW_ROOM"],
                    ),
                ),
            ],
        ]


follow_room_action = Action(
    name=FollowRoomAction.name,
    similes=FollowRoomAction().similes,
    description=FollowRoomAction.description,
    validate=FollowRoomAction().validate,
    handler=FollowRoomAction().handler,
    examples=FollowRoomAction().examples,
)

File: packages/typescript/src/features/advanced-capabilities/actions/followRoom.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { shouldFollowRoomTemplate } from "../../../prompts.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import {
	composePromptFromState,
	parseBooleanFromText,
	parseKeyValueXml,
} from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("FOLLOW_ROOM");
const FOLLOW_KEYWORDS = getValidationKeywordTerms("action.followRoom.request", {
	includeAllLocales: true,
});

export const followRoomAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	examples: (spec.examples ?? []) as ActionExample[][],
	validate: async (runtime: IAgentRuntime, message: Memory) => {
		const messageContentText = message.content.text;
		if (
			!messageContentText ||
			findKeywordTermMatch(messageContentText, FOLLOW_KEYWORDS) === undefined
		) {
			return false;
		}
		const roomId = message.roomId;
		const roomState = await runtime.getParticipantUserState(
			roomId,
			runtime.agentId,
		);
		return roomState !== "FOLLOWED" && roomState !== "MUTED";
	},
	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		_callback?: HandlerCallback,
		_responses?: Memory[],
	): Promise<ActionResult> => {
		if (!state) {
			logger.error(
				{
					src: "plugin:advanced-capabilities:action:follow_room",
					agentId: runtime.agentId,
				},
				"State is required for followRoomAction",
			);
			return {
				text: "State is required for follow room action",
				values: {
					success: false,
					error: "STATE_REQUIRED",
				},
				data: {
					actionName: "FOLLOW_ROOM",
					error: "State is required",
				},
				success: false,
				error: new Error("State is required for followRoomAction"),
			};
		}

		async function _shouldFollow(state: State): Promise<boolean> {
			const shouldFollowPrompt = composePromptFromState({
				state,
				template: shouldFollowRoomTemplate,
			});

			const response = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt: shouldFollowPrompt,
				stopSequences: [],
			});

			const parsed = parseKeyValueXml<{ decision?: boolean | string }>(
				response,
			);
			const decisionValue = parsed?.decision ?? response.trim();
			const cleanedResponse = String(decisionValue).trim().toLowerCase();

			if (
				parseBooleanFromText(decisionValue) ||
				cleanedResponse.includes("true") ||
				cleanedResponse.includes("yes")
			) {
				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							source: message.content.source,
							thought: "I will now follow this room and chime in",
							actions: ["FOLLOW_ROOM_STARTED"],
						},
					},
					"messages",
				);
				return true;
			}

			// Handle various negative responses
			if (
				cleanedResponse === "false" ||
				cleanedResponse === "no" ||
				cleanedResponse === "n" ||
				cleanedResponse.includes("false") ||
				cleanedResponse.includes("no")
			) {
				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							source: message.content.source,
							thought: "I decided to not follow this room",
							actions: ["FOLLOW_ROOM_FAILED"],
						},
					},
					"messages",
				);
				return false;
			}

			logger.warn(
				{
					src: "plugin:advanced-capabilities:action:follow_room",
					agentId: runtime.agentId,
					response,
				},
				"Unclear boolean response, defaulting to false",
			);
			return false;
		}

		const shouldFollow = await _shouldFollow(state);
		const room = state.data.room ?? (await runtime.getRoom(message.roomId));

		if (!room) {
			return {
				text: "Could not find room to follow",
				values: { success: false, error: "ROOM_NOT_FOUND" },
				data: { actionName: "FOLLOW_ROOM", error: "Room not found" },
				success: false,
			};
		}

		const roomName = room.name ?? `Room-${message.roomId.substring(0, 8)}`;

		if (shouldFollow) {
			try {
				await runtime.updateParticipantUserState(
					message.roomId,
					runtime.agentId,
					"FOLLOWED",
				);

				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							thought: `I followed the room ${roomName}`,
							actions: ["FOLLOW_ROOM_START"],
						},
					},
					"messages",
				);

				return {
					text: `Now following room: ${roomName}`,
					values: {
						success: true,
						roomFollowed: true,
						roomId: message.roomId,
						roomName: roomName,
						newState: "FOLLOWED",
					},
					data: {
						actionName: "FOLLOW_ROOM",
						roomId: message.roomId,
						roomName: roomName,
						followed: true,
					},
					success: true,
				};
			} catch (error) {
				logger.error(
					{
						src: "plugin:advanced-capabilities:action:follow_room",
						agentId: runtime.agentId,
						error: error instanceof Error ? error.message : String(error),
					},
					"Error following room",
				);
				return {
					text: "Failed to follow room",
					values: {
						success: false,
						error: "FOLLOW_FAILED",
					},
					data: {
						actionName: "FOLLOW_ROOM",
						error: error instanceof Error ? error.message : String(error),
						roomId: message.roomId,
					},
					success: false,
					error: error instanceof Error ? error : new Error(String(error)),
				};
			}
		} else {
			return {
				text: `Decided not to follow room: ${roomName}`,
				values: {
					success: true,
					roomFollowed: false,
					roomId: message.roomId,
					roomName: roomName,
					reason: "NOT_APPROPRIATE",
				},
				data: {
					actionName: "FOLLOW_ROOM",
					roomId: message.roomId,
					roomName: roomName,
					followed: false,
					reason: "Decision criteria not met",
				},
				success: true,
			};
		}
	},
};
